File: solutions/go/290.go
func wordPattern(pattern string, s string) bool {
	tokens := strings.Fields(s)

	if len(pattern) != len(tokens) {
		return false
	}

	h1 := make(map[byte]byte)
	h2 := make(map[string]byte)

	for i := 0; i < len(pattern); i++ {
		idx1, ok1 := h1[pattern[i]]
		idx2, ok2 := h2[tokens[i]]
		if ok1 != ok2 || idx1 != idx2 {
			return false
		}

		h1[pattern[i]] = byte(i)
		h2[tokens[i]] = byte(i)
	}

	return true
}

File: solutions/cpp/290.cc
class Solution {
public:
    bool wordPattern(string pattern, string s) {
        std::unordered_map<char, short> m1;
        std::unordered_map<std::string, short> m2;

        std::vector<std::string> tokens = split_whitespace(s);
        if (tokens.size() != pattern.size()) {
            return false;
        }

        for (short i = 0; i < pattern.size(); i++) {
            auto found1 = m1.find(pattern[i]);
            auto found2 = m2.find(tokens[i]);

            if (found1 == m1.end() && found2 == m2.end()) {
                m1[pattern[i]] = i;
                m2[tokens[i]] = i;
            } else if (found1 != m1.end() && found2 != m2.end()) {
                if (found1->second != found2->second) {
                    return false;
                }
            } else {
                return false;
            }
        }

        return true;
    }

private:
    std::vector<std::string> split_whitespace(std::string &s) {
        std::vector<std::string> result;
        std::size_t pos = 0;

        while (pos < s.length()) {
            auto found = s.find(' ', pos);
            if (found == std::string::npos) {
                result.push_back(s.substr(pos));
                break;
            } else {
                result.push_back(s.substr(pos, found - pos));
                pos = found + 1;
            }
        }

        return result;
    }
};
